[CUL-173] 사용자 프로필 수정 및 조회

euncheol-kim opened the pull request:
## 이슈 번호, 혹은 추가한 기능에 대해 설명해주세요.
[JIRA/CUL-173](https://chopinoff.atlassian.net/browse/CUL-173?atlOrigin=eyJpIjoiY2E0NDUzZmMyMDcwNDFmM2ExNjZiMWRjNGIyZTFlMGIiLCJwIjoiaiJ9)

## 작업 상세 내용

- [ ] 사용자 프로필 조회 (348e26fe2a46b2c48ac7c629356ff144c6f99853)

- [ ] 사용자 프로필 수정 (348e26fe2a46b2c48ac7c629356ff144c6f99853)
  - [ ] 어노테이션 기반 nickname 중복 검증 기능 추가 (348e26fe2a46b2c48ac7c629356ff144c6f99853)

- [ ] User 필드 추가 및 수정 (4e7b7cd8c7914005830fd5640baf39a2d36e3ec7)
  - [ ] password 필드 삭제 : 소셜 로그인 기반 서비스이므로 필요없는 필드라 판단하여 삭제 진행
  - [ ] preferredCategory 추가 : 사용자의 선호하는 카테고리 컬럼 작성 -> 데이터베이스의 JSON 타입으로 매핑

- [ ] GlobalException 코드 작성 (e3cb4106d1eadb5ca277ae8bf71e69ab8e142207)

## 기타 내용

euncheol-kim committed:
[fix🐛] 패키지 명 수정

euncheol-kim committed:
Merge branch 'dev' of https://github.com/CultureSpot/CultureSpot-Backend into feature/#CUL-173/user

euncheol-kim committed:
Merge branch 'dev' of https://github.com/CultureSpot/CultureSpot-Backend into feature/#CUL-173/user

euncheol-kim committed:
[feat] User 엔티티 수정 (CUL-173)

# 대표 수정내역
# 1. password 필드 삭제 : 현재 서비스는 소셜 로그인만 지원하기 때문에 삭제 진행
# 2. preferredCategory 필드 추가 : 사용자의 선호하는 카테고리
# 3. updatePreferredCategory(List<String> categories) : 필드 상태 변경을 위한 도메인 메소드 생성

euncheol-kim committed:
[feat] 사용자 프로필 조회 및 수정 기능 개발 (CUL-173)

# 대표 개발 내역
# 1. 검증 : 어노테이션 기반 닉네임 중복에 대한 검증 (ValidNicknameValidator, ValidNickname)
# 2. 응답 : FE에서 응답 데이터를 한 번 감싸는 형태로 받기를 원하여 ApiResponse, NamedWrapper 추가 작업 진행
# 3. 사용자 프로필 GET 조회 및 PUT 수정 기능 개발

euncheol-kim committed:
[feat] 사용자의 잘못된 권한에 대한 코드 추가 (CUL-173)

euncheol-kim committed:
[feat] 카테고리를 문자열 요청으로 받았을 때 name값과 비교하여 Category 반환하는 메소드 추가  (CUL-173)

euncheol-kim committed:
[feat] 현재 로그인한 사용자의 정보를 가져오는 프로바이더 추가  (CUL-173)

euncheol-kim committed:
[refactor] 사용자 생성과 관련한 로직을 userService내부로 캡슐화  (CUL-173)

euncheol-kim committed:
[refactor] 사용자 패스워드 컬럼을 삭제함에 따른 코드 수정  (CUL-173)

euncheol-kim committed:
[feat] MethodArgumentNotValidException에 대한 에러 핸들링 코드 작성 (CUL-173)

euncheol-kim committed:
[comment] 필요없는 주석 삭제(CUL-173)

euncheol-kim committed:
[comment] 사용자 생성 메소드명 변경에 따른 수정(CUL-173)

euncheol-kim committed:
[feat] @PreAuthorize사용을 하기위해 @EnableMethodSecurity추가(CUL-173)

euncheol-kim committed:
[feat] validation 의존성 추가(CUL-173)

euncheol-kim committed:
[rename] 사용자 정보 수정 요청에 대한 메소드명 변경(CUL-173)

euncheol-kim committed:
[refactor] NamedWrapper 제네릭 추가(CUL-173)

# 리팩토링 의의
# - Swagger에서 응답값을 만들 때 범용성 있게 사용하기 위함
# - 대표 적용 케이스 : SwaggerUserProfileResponse

euncheol-kim committed:
[feat] Swagger 설정 진행 (CUL-173)

euncheol-kim committed:
[docs] 사용자 프로필 조회/수정에 대한 Swagger 적용(CUL-173)

euncheol-kim committed:
[remove] 파일 삭제 진행(CUL-173)

# 삭제 의의
# - 사용하지 않음

euncheol-kim committed:
[refactor] 사용자 정보 조회 및 수정 기능 리팩토링 진행(CUL-173)

# 리팩토링 이유
# - 자체적으로 추가한 ApiResponse와 NamedWrapper를 사용하지 않아도 될 것 같아서 수정 진행
# - ApiResponse와 NamedWrapper는 차후 사용할 수 있을 것으로 생각하여 유지 (삭제요청시 삭제하겠습니다)
# 리팩토링 내역
# 1. 사용자 정보 조회 반환값 수정 (UserController.getUserProfile(@Auth User user))
# 2. UserProfile dto 생성 : Json response에서 user로 한 번 감싸주기 위함
# 3. 코드 변경에 따른 UserMapper의 수정
# 4. 응답 예시 리스폰스 삭제 진행

euncheol-kim committed:
[rename] global 패키지 내부에 존재하던 파일들 단순 경로 수정(CUL-173)

# 수정의의
# Domain모듈을 보면 core와 같은 레벨에 common 패키지가 존재하는데, 일관성을 맞춰주고자 패키지 경로 수정
# 이로인해 core패키지는 도메인만 남아 책임과 역할 부분에 있어서 더 명확해진다고 생각합니다.
# etc. Domain 모듈의 경우에도 core하위에 global이 존재하는데 이것도 차후에 옮기면 어떨까 고민중입니다.

spearkkk commented on CultureSpot-Domain/src/main/java/com/culturespot/culturespotdomain/core/user/entity/User.java:
```diff
@@ -24,24 +26,27 @@ public class User extends BaseEntity {
 
-    @OneToMany(mappedBy = "user", fetch = FetchType.EAGER, cascade = CascadeType.ALL)
+    @OneToMany(mappedBy = "user", fetch = FetchType.LAZY, cascade = CascadeType.ALL, orphanRemoval = true)
     private Set<UserRole> roles = new HashSet<>();
```
UserRole이 entity네요.
사실 UserRoleType이 UserRole의 전부라서
UserRole이라는 entity가 있을 이유를 잘 모르겠어요.

UserRoleType을 Set<UserRoleType>으로도 충분한 걸로 보여서요.

한번 고민해보면 좋을 듯 합니다.

euncheol-kim replied:
해당 부분은 초기에 작업하면서 `사용자가 여러 권한을 갖게하면 좋겠다.`라고 생각하고 작업 한 부분입니다. 그러다보니 테이블의 관계에서 `USER-1-N-USER_ROLE-N-1-ROLE`이라는 관계를 성립하고 엔티티로 작업을 해버리게 됐습니다. 


혹시 이 부분은 팀 적으로 고민하면서 코드 변경을 이루어도 괜찮을까요?

spearkkk replied:
그러면 질문을 바꿔서 해보면,
여기서의  ROLE은 어떤 값들이 있나요?

euncheol-kim replied:
ROLE은 ADMIN과 USER만 있습니다.

spearkkk replied:
사용자가 여러 권한을 가지면 좋겠다 -> user: role = 1:N 으로 보여서요.
N:1은 어디서 나온거죠?

euncheol-kim replied:
현재 구현된 기준으로 말씀드리겠습니다.

Role의 경우 `서비스에서 제공하는 권한에 대한 관리 테이블`로 생각했습니다.

따라서
(1) 한 명의 사용자는 여러 권한을 가질 수 있다.
(2) 하나의 권한은 여러 사용자가 가질 수 있다. ( = 각 권한은 여러 사용자가 가질 수 있다)
라고 생각하여 N:M으로 생각했고 중간테이블 `USER_ROLE`을 만들어야 한다고 생각했었습니다.

혹시 코멘트에 대한 답변이 충분히 되었는지 여쭤봐도 될까요?

spearkkk replied:
>Role의 경우 서비스에서 제공하는 권한에 대한 관리 테이블로 생각

라고 해주셨느데요. 그러면 role이라는 데이터를 저희 도메인의 데이터로 봐야하는데,
말해주셨듯이 USER와 ADMIN밖에 없습니다.
이를 데이터베이스의 테이블로 관리한다면 얻을 수 있는 장점이 있을까요?

그렇다면, Category, EventType, 또는 성별, 나이대, 등의 메타데이터는 어떻게 관리하면 좋을까요?

euncheol-kim replied:
> 데이터베이스의 테이블로 관리한다면 얻을 수 있는 장점이 있을까요?

사실 없을 것 같습니다.

> 그렇다면, Category, EventType, 또는 성별, 나이대, 등의 메타데이터는 어떻게 관리하면 좋을까요?

아.. 앞서 말씀해주신 `Category`, `EventType`은 코드레벨의 열거형 객체로 관리하고 있기 때문에 데이터베이스에서 관리하는 테이블로 빼지 않아도 되고 그럴 필요 조차 없다고 이해하면 되는 부분일까요?

euncheol-kim replied:
말씀해주신 부분을 하나 하나 생각해보니 수정이 필요한 부분은 틀림없을 것 같습니다..!

spearkkk replied:
사실, 이런 값들을 마스터 코드 자체로 테이블로 관리하기도 합니다.

---

> (2) 하나의 권한은 여러 사용자가 가질 수 있다

권한을 값 자체로 보지 않고 엔티티로 봐서 나온 생각으로 보여지는데요.
권한을 엔티티로 보는 이유가 있을까요?

euncheol-kim replied:
엔티티로 봐야하는 이유에 대해서는 충분히 생각해보지는 못했습니다. 이는 경험의 부족에서 나오는 것 같습니다.

지금까지 상태에 대해서 관리를 한다면 상태는 반드시 테이블에 넣어주고 작업을 해야한다는 잘못된 생각을 가지고 있었던것 같습니다..!

더불어, 작업해주신 Category 부분과 대조해보니 상태값이라고 꼭 테이블로 관리할 필요는 없다는 것을 배웠습니다. 감사합니다..!

해당 부분의 수정이 이루어져야 할 것 같은데 사용자가 다양한 권한을 가진다는 롤이 있다면 json타입으로 만드는게 적절할까요?

spearkkk replied:
넵~ 문자열로 관리해도 나쁘지 않아보입니다~

spearkkk commented on CultureSpot-Domain/src/main/java/com/culturespot/culturespotdomain/core/user/entity/User.java:
```diff
@@ -50,21 +55,29 @@ public User(
             int profileCode,
-            LocalDateTime lastLoginAt
+            LocalDateTime lastLoginAt,
+            Map<Category, String> preferredCategory
```
Map<Category, String> 인 이유를 모르겠습니다.

euncheol-kim replied:
![image](https://github.com/user-attachments/assets/966f5f06-6e47-47a6-b108-e1a952be859d)
사용자가 선호하는 카테고리는 이미지의 32행 요청으로 받게 됩니다.


![image](https://github.com/user-attachments/assets/0eed1101-c3f0-4fdd-a5f9-6a7e39c27b58)
![image](https://github.com/user-attachments/assets/5faf9774-343f-4a7b-b7db-dae04c897cd8)

선호하는 카테고리를 데이터베이스에 Json타입으로 저장하기로 결정을 하고, KEY와 VALUE에 대한 기준을 어떻게 해야할지 고민을 했는데요. 생각하길, 상수 그 자체의 이름을 Key로 (`Music`, `ART`, `ETC` 등), Category에 정의된 name을 value로 가져가면 좋겠다고 생각하여 작업한 내역입니다. 

사실 Key와 Value를 결정하는 부분에 있어서 한 번 여쭤보긴 했어야하는데 그러지 못해서 죄송합니다. 혹시 다른 방식을 말씀해주시면 바꿔보도록하겠습니다.

spearkkk replied:
이미 Enum인 Category에 value 값이 다 있는데, 디비에 왜 저장하는지 모르겠어요.

euncheol-kim replied:
무슨 말씀이신지 이해했습니다. 짚어주셔서 감사합니다.
그러면 저장값을 어떻게 가져가는게 좋을까요?

```
{
  "select": ["ETC", "MUSIC", "VIDEO"]
}
```

지금 당장 떠오르는 건 위와 같은 형태인데 위의 형태처럼 저장해야할까요?

spearkkk replied:
넵 저는 enum값만 저장해도 충분하다고 생각합니다.

euncheol-kim replied:
네 코멘트 반영하도록하겠습니다. 감사합니다.

spearkkk commented on CultureSpot-Domain/src/main/java/com/culturespot/culturespotdomain/core/user/service/UserServiceImpl.java:
```diff
@@ -26,50 +33,46 @@ public User findUserOrThrow(String email) {
                 .orElseGet(() -> {
-                    User newUser = createUser(email, authType);
-                    return userRepository.save(newUser);  // 저장 후 반환
+                    User user = createUser(email, authType);
```
아래는 @Transactional 로 감쌌는데, 왜 여기는 없을까요?

euncheol-kim replied:
헉 빠진 것 같습니다. 재수정하겠습니다. 감사합니다.

euncheol-kim replied:
Resolve 타겟 커밋 : [44e7f74](https://github.com/culturespot/culturespot-backend/pull/59/commits/44e7f746bcaf31f225c9ad16fd274c0593b819be)

spearkkk commented on CultureSpot-ServiceApi/src/main/java/com/culturespot/culturespotserviceapi/core/user/controller/spec/description/UserControllerDescription.java:
```diff
@@ -0,0 +1,26 @@
+package com.culturespot.culturespotserviceapi.core.user.controller.spec.description;
+
+public class UserControllerDescription {
```
UserControllerSpec 이 안에서만 쓰는거면 안으로 넣을 수 있지 않나요?

euncheol-kim replied:
![image](https://github.com/user-attachments/assets/d14ae950-4265-4cf4-9306-f7e2ee723c5f)

말씀해주신 안으로 넣는다는 의미가 `UserControllerSpec` 내부에 작성해주면 좋을 것 같다는 말씀으로 이해가 됩니다.  해당 파일이 사용중인 곳은 이미지의 24행 31행인데요. 해당 내용을 여기다가 쓰기에는 코드 줄이 너무 길어질 것 같아 분리하는 것이 좋다고 생각해 파일을 추가한 부분입니다!

좋지 못한 설계일까요?

spearkkk replied:
Description랑 Spec이랑 긴밀하게 연관이 있고 다른 클래스에서는 description 클래스를 아예 사용하지 않을 것 같아요.
긴밀하게 연결되어 있는 부분이 두개의 별도로 나눠져 있는건 괜찮을까요?

euncheol-kim replied:
사실 파일을 너무 쪼개서 관리를 하다보면 프로젝트가 성숙해짐에 따라 복잡도가 높아지고, 비슷한 개발건에서 이런식의 일관성있는 개발이 이루어질 경우 문제가 발생할 것이라 생각하게 됩니다. (사실 지금은 제 코딩 스타일이 많이 반영된 상태이기도 합니다..)

spearkkk replied:
개인적인 스타일이면 괜찮습니다~

euncheol-kim replied:
실무에서는 어떤 권장 방향을 선호하는지 여쭤봐도 될까요? 권장해주시는 방향으로 적용하고 싶습니다.

spearkkk replied:
편하신 방법이 더 맞을 것 같아요.
실무에서는 누군가가 불편하면 얘기하고 바꾸는 거라서요. 여기서는 은철님 의견이 더 맞아보여요~

euncheol-kim replied:
아 네 알겠습니다. 해당 건은 resolve하겠습니다. 감사합니다.

euncheol-kim committed:
[refactor] 사용자 등록 메소드에 @Transactional 적용

euncheol-kim committed:
[refactor] 사용자의 선호하는 카테고리의 타입 변경과 그에 따른 수정 (CUL-270)

# 주요 수정안: User.preferredCategory
#  - 타입변경: Map<Category, String> to Map<String, List<Category>>
#  - 변경의의: 데이터베이스에 단순히 enum 값을 저장하기 위함 (기존에는 enum과 값을 같이 저장하는 비효율적 방식)

euncheol-kim committed:
Merge branch 'dev' of https://github.com/CultureSpot/CultureSpot-Backend into feature/CUL-173/user
